From 07a7eefbbdc6346ad2fa98c723c4630bac6474d9 Mon Sep 17 00:00:00 2001
From: "Simental Magana, Marcos" <marcos.simental.magana@intel.com>
Date: Tue, 29 Sep 2015 10:12:24 -0500
Subject: [PATCH] Integrate OSprofiler in Keystone

-Add settings for OSprofiler wsgi middleware, This middleware is used
for 2 things:
1) It checks that person who want to trace is trusted and knows
secret HMAC key (that is specified in paste.ini).
2) It initalize profiler in case of proper trace headers
and add first wsgi trace point, with info about HTTP request.

*) Init profiler on start in both cases (httpd or eventlet)
*) Adding new conf group for profiler (to enable/disable)
*) Adding trace point for all DB (sql) calls
*) By default osprofiler is disabled

To test:

$ . amdminrc # you should be admin user/tenant to fetch profiling data
$ keystone --profile SECRET_KEY user-list
$ osprofiler trace show --html --out result.html <TRACE_ID>
---
 etc/keystone-paste.ini      |  9 +++++++--
 keystone/common/config.py   |  6 ++++++
 keystone/common/profiler.py | 41 +++++++++++++++++++++++++++++++++++++++++
 keystone/common/sql/core.py |  4 ++++
 keystone/notifications.py   |  5 +++++
 keystone/server/eventlet.py |  4 ++++
 keystone/server/wsgi.py     |  3 +++
 requirements.txt            |  1 +
 8 files changed, 71 insertions(+), 2 deletions(-)
 create mode 100644 keystone/common/profiler.py

diff --git a/etc/keystone-paste.ini b/etc/keystone-paste.ini
index 70db382..b6baaad 100644
--- a/etc/keystone-paste.ini
+++ b/etc/keystone-paste.ini
@@ -85,10 +85,10 @@ use = egg:keystone#public_version_service
 use = egg:keystone#admin_version_service
 
 [pipeline:public_version_api]
-pipeline = sizelimit url_normalize public_version_service
+pipeline = sizelimit osprofiler url_normalize public_version_service
 
 [pipeline:admin_version_api]
-pipeline = sizelimit url_normalize admin_version_service
+pipeline = sizelimit osprofiler url_normalize admin_version_service
 
 [composite:main]
 use = egg:Paste#urlmap
@@ -101,3 +101,8 @@ use = egg:Paste#urlmap
 /v2.0 = admin_api
 /v3 = api_v3
 / = admin_version_api
+
+[filter:osprofiler]
+paste.filter_factory = osprofiler.web:WsgiMiddleware.factory
+hmac_keys = SECRET_KEY
+enabled = yes 
diff --git a/keystone/common/config.py b/keystone/common/config.py
index 107fbbf..e5d3538 100644
--- a/keystone/common/config.py
+++ b/keystone/common/config.py
@@ -1165,6 +1165,12 @@ FILE_OPTIONS = {
                     deprecated_for_removal=True,
                     help='Require client certificate.'),
     ],
+    'profiler': [
+        cfg.BoolOpt("enabled", default=False,
+                    help='If False fully disable profiling feature.'),
+        cfg.BoolOpt("trace_sqlalchemy", default=False,
+                    help="If False doesn't trace SQL requests.")
+    ],
 }
 
 
diff --git a/keystone/common/profiler.py b/keystone/common/profiler.py
new file mode 100644
index 0000000..b4c1357
--- /dev/null
+++ b/keystone/common/profiler.py
@@ -0,0 +1,41 @@
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+import logging
+
+import osprofiler.notifier
+import oslo_messaging
+
+from keystone import config
+from keystone import notifications
+
+
+CONF = config.CONF
+LOG = logging.getLogger(__name__)
+
+
+def setup(name, host='0.0.0.0'):
+
+    if CONF.profiler.enabled:
+        _notifier = osprofiler.notifier.create(
+            "Messaging", oslo_messaging, {},
+            notifications.get_transport(), "keystone", name, host)
+        osprofiler.notifier.set(_notifier)
+        LOG.info("OSProfiler is enabled.\nIt means that person who knows "
+                 "any of hmac_keys that are specified in "
+                 "/etc/keystone-paste.ini can trace his requests. \n"
+                 "In real life only operator can read this file so there "
+                 "is no security issue. Note that even if person can "
+                 "trigger profiler, only admin user can retrieve trace "
+                 "information.\n"
+                 "To disable OSprofiler set in keystone.conf:\n"
+                 "[profiler]\nenabled=false")
diff --git a/keystone/common/sql/core.py b/keystone/common/sql/core.py
index ebd61bb..8f89a4d 100644
--- a/keystone/common/sql/core.py
+++ b/keystone/common/sql/core.py
@@ -28,6 +28,7 @@ from oslo_db.sqlalchemy import models
 from oslo_db.sqlalchemy import session as db_session
 from oslo_log import log
 from oslo_serialization import jsonutils
+import osprofiler.sqlalchemy
 import six
 import sqlalchemy as sql
 from sqlalchemy.ext import declarative
@@ -174,6 +175,9 @@ def _get_engine_facade():
 
     if not _engine_facade:
         _engine_facade = db_session.EngineFacade.from_config(CONF)
+        if CONF.profiler.enabled and CONF.profiler.trace_sqlalchemy:
+            osprofiler.sqlalchemy.add_tracing(sql, _engine_facade.get_engine(),
+                                              "db")
 
     return _engine_facade
 
diff --git a/keystone/notifications.py b/keystone/notifications.py
index bea09d3..65fed68 100644
--- a/keystone/notifications.py
+++ b/keystone/notifications.py
@@ -359,6 +359,11 @@ def notify_event_callbacks(service, resource_type, operation, payload):
                 cb(service, resource_type, operation, payload)
 
 
+def get_transport():
+    """Return transport object."""
+    return oslo_messaging.get_transport(CONF)
+
+
 def _get_notifier():
     """Return a notifier object.
 
diff --git a/keystone/server/eventlet.py b/keystone/server/eventlet.py
index 243f023..d2707ff 100644
--- a/keystone/server/eventlet.py
+++ b/keystone/server/eventlet.py
@@ -33,6 +33,7 @@ oslo_i18n.enable_lazy()
 
 
 from keystone.common import environment
+from keystone.common import profiler
 from keystone.common import utils
 from keystone import config
 from keystone.i18n import _
@@ -65,6 +66,9 @@ def create_server(conf, name, host, port, workers):
     server = environment.Server(app, host=host, port=port,
                                 keepalive=CONF.eventlet_server.tcp_keepalive,
                                 keepidle=CONF.eventlet_server.tcp_keepidle)
+
+    profiler.setup(name, host)
+
     if CONF.eventlet_server_ssl.enable:
         server.set_ssl(CONF.eventlet_server_ssl.certfile,
                        CONF.eventlet_server_ssl.keyfile,
diff --git a/keystone/server/wsgi.py b/keystone/server/wsgi.py
index ae24c48..e84fbe9 100644
--- a/keystone/server/wsgi.py
+++ b/keystone/server/wsgi.py
@@ -27,6 +27,7 @@ oslo_i18n.enable_lazy()
 
 
 from keystone.common import environment
+from keystone.common import profiler
 from keystone import config
 import keystone.middleware.core as middleware_core
 from keystone.server import common
@@ -63,6 +64,8 @@ def initialize_application(name):
     _unused, application = common.setup_backends(
         startup_application_fn=loadapp)
 
+    profiler.setup(name, CONF.eventlet_server.public_bind_host)
+
     # Create a CORS wrapper, and attach keystone-specific defaults that must be
     # included in all CORS responses
     application = cors.CORS(application, CONF)
diff --git a/requirements.txt b/requirements.txt
index 555a148..ab6e6e8 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -36,3 +36,4 @@ dogpile.cache>=0.5.4
 jsonschema!=2.5.0,<3.0.0,>=2.0.0
 pycadf>=1.1.0
 msgpack-python>=0.4.0
+osprofiler>=0.3.0
-- 
2.4.3

